backend/push/client.go

```diff
@@ -23,36 +27,130 @@ func NewService(ctx context.Context, firebaseProjectID string) (*Service, error)
 	}
 	return &Service{
 		app,
+		queries,
 	}, nil
 }
 
-// SendPushNotification sends a push notification to the specified tokens
-func (s *Service) SendPushNotification(ctx context.Context, tokens []string) {
-	// TODO: actually implement notifications, not this placeholder things
+func notificationToPayload(notification common.Notification) map[string]string {
+	var payload = map[string]string{}
+	for lan, val := range notification.Title {
+		if val.Valid {
+			payload["title_"+lan] = val.ValueOrZero()
+		}
+	}
+	for lan, val := range notification.Description {
+		if val.Valid {
+			payload["description_"+lan] = val.ValueOrZero()
+		}
+	}
+	for lan, val := range notification.Images {
+		if val.Valid {
+			payload["image_"+lan] = val.String
+		}
+	}
+	return payload
+}
+
+func (s *Service) pushMessages(ctx context.Context, messages []*messaging.Message) error {
 	client, err := s.app.Messaging(ctx)
 	if err != nil {
-		return
+		return err
 	}
 
+	const maxConcurrent = 200
+
+	var ranges [][]*messaging.Message
+
+	for i := 0; i < len(messages); i += maxConcurrent {
+		var r int
+		if i+maxConcurrent > len(messages) {
+			r = len(messages)
+		} else {
+			r = i + maxConcurrent
+		}
+		if r == 0 {
+			break
+		}
+		ranges = append(ranges, messages[i:r])
+	}
+
+	errors := parallel.Map(ranges, func(r []*messaging.Message, _ int) []error {
+		res, err := client.SendAll(ctx, r)
+		//TODO: Implement error handling
+		if err != nil {
+			return []error{err}
+		}
+		return lo.Map(lo.Filter(res.Responses, func(r *messaging.SendResponse, _ int) bool {
+			return r.Error != nil
+		}), func(r *messaging.SendResponse, _ int) error {
+			return r.Error
+		})
```

Comment:
Maybe a little comment here as to what is going on :) High energy lines of code.